homeassistant/components/nuki/__init__.py

```diff
@@ -49,11 +50,12 @@ def _update_devices(hass: HomeAssistant, devices: list[NukiDevice]) -> None:
                     device.update(level)
 
                     if not last_ring_action_state and device.ring_action_state:
+                        entity_id = ent_reg.async_get_entity_id(Platform.LOCK, DOMAIN, device.nuki_id) 
```

Comment:
Preferably we shouldn't use the async API in sync context, although technically in this case it works. Not sure what's a better alternative at the moment.

Reply:
Didn't see a sync alternative. As the `async_fire` specifically states that it shouldn't be used outside the event loop, I thought this one would work. Anyway, the `async_fire` worked too, so I'm not sure. I found nearly no documentation about it

Reply:
It depends what the call does. This case is just a dict lookup which is canonical. But it's bad to use the async API in sync context regardless as we may change details about how it works in the future.

Reply:
I'll modify the _update_devices to be async, and call the device.update()  with `hass.async_add_executor_job()`. I'm not sure how much overhead may `run_in_executor` add inside that loop though

Reply:
There's an overhead to switch context. That's why we have this function so we only need to schedule it once on the executor for all devices. Compare with needing to schedule on the executor for each device call.

Reply:
Comparing times, it doesn't seem to be relevant. With the 2 devices I have:
|  | 1 run_in_executor | 1 run_in_executor per device |
| ------------- | ------------- | ------------- |
| First fetch | 0.6s | 0.8s  |
| Other fetches | 0.05s | 0.05s |
(Those are just approximate measures as it's very variable)

Other way would be to return from that function a `dict[event_type, set[unique_id]]`. Not sure if it's worth it, but I'll see how the code ends up

Reply:
Moved the event firing logic outside that function, I think it's pretty clear now